base/mmap.jl

```diff
@@ -1,168 +1,177 @@
+settings(io::IO) = Ptr{Cwchar_t}(C_NULL), isreadonly(io), true
+end # @windows_only
 
-@windows_only type SharedMemSpec
-    name :: AbstractString
-    readonly :: Bool
-    create :: Bool
+# core impelementation of mmap
+# we define a dummy Array type to achieve the Mmap.Array interface
+type Array{T,N}
+end
+type BitArray{N}
 end
 
-@windows_only begin
-# Mmapped-array constructor
-function mmap_array{T,N}(::Type{T}, dims::NTuple{N,Integer}, s::Union(IO,SharedMemSpec), offset::FileOffset)
-    if isa(s,IO)
-        hdl = _get_osfhandle(RawFD(fd(s))).handle
-        if Int(hdl) == -1
-            error("could not get handle for file to map: $(FormatMessage())")
-        end
-        name = Ptr{Cwchar_t}(C_NULL)
-        ro = isreadonly(s)
-        create = true
-    else
-        # shared memory
-        hdl = -1
-        name = utf16(s.name)
-        ro = s.readonly
-        create = s.create
-    end
-    len = prod(dims)*sizeof(T)
-    const granularity::Int = ccall(:jl_getallocationgranularity, Clong, ())
-    if len < 0
-        throw(ArgumentError("requested size must be ≥ 0, got $len"))
-    end
-    if len > typemax(Int)-granularity
-        throw(ArgumentError("file is too large ot memory-map on this platform"))
-    end
-    # Set the offset to a page boundary
-    offset_page::FileOffset = div(offset, granularity)*granularity
-    szfile = convert(Csize_t, len + offset)
-    szarray = szfile - convert(Csize_t, offset_page)
-    access = ro ? 4 : 2
-    if create
-        flprotect = ro ? 0x02 : 0x04
-        mmaphandle = ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
-            hdl, C_NULL, flprotect, szfile>>32, szfile&typemax(UInt32), name)
-    else
-        mmaphandle = ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-            access, true, name)
-    end
-    if mmaphandle == C_NULL
-        error("could not create file mapping: $(FormatMessage())")
-    end
-    viewhandle = ccall(:MapViewOfFile, stdcall, Ptr{Void}, (Ptr{Void}, Cint, Cint, Cint, Csize_t),
-        mmaphandle, access, offset_page>>32, offset_page&typemax(UInt32), szarray)
-    if viewhandle == C_NULL
-        error("could not create mapping view: $(FormatMessage())")
-    end
-    A = pointer_to_array(convert(Ptr{T}, viewhandle+offset-offset_page), dims)
-    finalizer(A, x->Libc.munmap(viewhandle, mmaphandle))
+function Mmap.Array{T,N}(::Type{T}, dims::NTuple{N,Integer}, io::IO, offset::Integer=position(io); grow::Bool=true, shared::Bool=true)
+    # check inputs
+    isopen(io) || throw(ArgumentError("$io must be open to mmap"))
+
+    len = prod(dims) * sizeof(T)
+    len > 0 || throw(ArgumentError("requested size must be > 0, got $len"))
+    len < typemax(Int) - PAGESIZE || throw(ArgumentError("requested size must be < $(typemax(Int)-PAGESIZE), got $len"))
+
+    offset >= 0 || throw(ArgumentError("requested offset must be ≥ 0, got $offset"))
+
```

Comment:
I think this should also check `isbits(T)`, since mmapping an array of pointers doesn't make much sense and would probably make Julia segfault.